homeassistant/components/bluetooth_le_tracker/device_tracker.py

```diff
@@ -117,13 +139,41 @@ def discover_ble_devices():
     def update_ble(now):
         """Lookup Bluetooth LE devices and update status."""
         devs = discover_ble_devices()
+        if devs_track_battery:
+            adapter = hass.data[DATA_BLE][DATA_BLE_ADAPTER]
         for mac in devs_to_track:
             if mac not in devs:
                 continue
 
             if devs[mac] is None:
                 devs[mac] = mac
-            see_device(mac, devs[mac])
+
+            battery = None
+            if (
+                mac in devs_track_battery
+                and now > devs_track_battery[mac] + battery_track_interval
+            ):
+                handle = None
+                try:
+                    adapter.start(reset_on_start=True)
+                    _LOGGER.debug("Reading battery for Bluetooth LE device %s", mac)
+                    bt_device = adapter.connect(mac)
+                    # Try to get the handle; it will raise a BLEError exception if not available
+                    handle = bt_device.get_handle(BATTERY_CHARACTERISTIC_UUID)
+                    battery = ord(bt_device.char_read(BATTERY_CHARACTERISTIC_UUID))
+                    devs_track_battery[mac] = now
+                except pygatt.exceptions.NotificationTimeout:
+                    _LOGGER.warning("Timeout when trying to get battery status")
+                except pygatt.exceptions.BLEError as err:
+                    _LOGGER.warning("Could not read battery status: %s", err)
+                    if handle is not None:
+                        # If the device does not offer battery information, there is no point in asking again later on.
+                        # Remove the device from the battery-tracked devices, so that their battery is not wasted
+                        # trying to get an unavailable information.
+                        del devs_track_battery[mac]
+                finally:
+                    adapter.stop()
```

Comment:
Can this throw an exception?

Reply:
pygatt doesn't throw any exceptions on "stop". From what I could follow of the calls inside the method, it doesn't look as if it would throw any exception also from deeper in the implementation.